New page on crossplane pods

plumbis opened the pull request:
Creates a new concept page covering the crossplane core pod and rbac manager pod.

jeanduplessis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+## Crossplane pod
+
+### Init pod
+Before starting the core Crossplane pod an _init_ Pod runs. The init pod is
```
Why is _init_ `P`od written with a capital here but in lowercase almost everywhere else?

bobh66 replied:
This should be "the init container" runs.  The Pod has "initContainers" that run at startup and "containers" that run after startup.

plumbis replied:
fixed

jeanduplessis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+## Crossplane pod
+
+### Init pod
+Before starting the core Crossplane pod an _init_ Pod runs. The init pod is
```
Why is _init_ `P`od written with a capital here but in lowercase almost everywhere else?

jeanduplessis replied:
Why is _init_ `P`od written with a capital here but in lowercase almost everywhere else?

jeanduplessis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+description: Background on the components installed with Crossplane and their functions.
+---
+
+The base Crossplane installation consists of two Pods, the `crossplane` pod and
```
Should two `P`ods be capitalized here?

plumbis replied:
fixed.

jeanduplessis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+information about customizing the Crossplane install.
+{{< /hint >}}
+
+The init pod completes and starts the Crossplane core pod. 
```
What happens to the init pod after completion? Does the end user have to care about it at all after completion? Does it also run in the `crossplane-system` namespace?

plumbis replied:
I added more information about viewing the init container in `k get pods` and that starts the core container automatically.

jeanduplessis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+
+### Core pod
+
+The main Crossplane pod, called the _core_ pod, enforces
```
It's getting confusing with the naming here. We refer to the main Crossplane pod as "core" pod, but its name is `crossplane` in the `crossplane-system` namespace, right? Not sure we can do any better than what's written but just calling it out.

plumbis replied:
More accurately it's the init _container_ and core _container_. Both of these containers are in the `crossplane` _pod_ inside the `crossplane-system` namespace. The name "core" is what Crossplane calls it in source. 

Would being more specific that they are two _containers_ inside a single _pod_ help? Otherwise we're stuck with

jeanduplessis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+
+For objects that Kubernetes can't provide a watch for, Crossplane
+periodically poll the resource to find it's state. The default polling rate is
+one minute. Change the polling rate with the `---poll-interval` pod argument.
```
Where should one change the polling rate? Can we be more specific than just stating to add it as a pod argument. Also is there any considerations that need to be taken into account when changing the poll-interval?

bobh66 replied:
The polling interval is an input to the helm chart for Crossplane deployment

plumbis replied:
I added some text on the impact of changing the poll-interval. 

@jeanduplessis are you okay with the note on "Read the Change Pod Settings section...." for information on applying the change or do you want more here?

jeanduplessis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+When installing a Provider their service account needs permissions to create and 
+manage managed resources. 
+
+When creating Composite Resources the Crossplane service account needs
```
Why do we capitalize Composite Resources here, yet above we write managed resources in lowercase?

plumbis replied:
fixed

jeanduplessis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+When installing a Provider their service account needs permissions to create and 
+manage managed resources. 
+
+When creating Composite Resources the Crossplane service account needs
```
Why do we capitalize Composite Resources here, yet above we write managed resources in lowercase?

jeanduplessis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+* `All` - The default policy, configures and manages all RBAC capabilities.
+* `Basic` - Only manage policies related to Composite resources. 
+
+Change from the default `All` to the `Basic` policy with the `-m Basic` pod 
```
Why/when would one want to switch from All to Basic policy?

jeanduplessis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+Change from the default `All` to the `Basic` policy with the `-m Basic` pod 
+argument.
+
+Disable the RBAC manager after installation by deleting the
+`crossplane-rbac-manager` deployment from the `crossplane-system` namespace.
+
+Disable the RBAC manager before installation by editing the Helm `values.yaml`
+file, setting `rbacManager.deploy` to `false`.
```
Under what circumstances would you want to disable the RBAC manager?

plumbis replied:
Added a note.

jeanduplessis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+
+## Leader election
+
+By default only a single Crossplane pod runs in a cluster. If more than one
```
Why/how would you run more than one Crossplane pod in a cluster?

plumbis replied:
Added a note that you can run more than one but most people don't need it. (Feedback from Nic)

jeanduplessis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+Leader election applies to both the Crossplane pod and the RBAC manager pod.
+{{< /hint >}}
+
+Enable leader elections with the `--leader-election` pod argument.
```
Would this argument set which XP pod becomes the leader?

bobh66 replied:
No - it is random based on timing.  It is even possible that the "failed" pod could regain the leader role after it restarts if the timer for the "standby" pod(s) has not expired yet.

plumbis replied:
No, "who" is the leader doesn't matter since it's just another pod in the deployment. Just like the specific node a pod runs on doesn't matter. 

The election itself is driven by kubernetes. 

Should I add more context?

jeanduplessis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+
+## Change pod settings
+
+Changing Crossplane pod settings either before installing Crossplane by editing
```
"Changing Crossplane pod..." should be "Change Crossplane pod..."

plumbis replied:
fixed

jeanduplessis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+{{< /hint >}}
+
+Add Crossplane pod arguments to the 
+{{<hover label="args" line="7" >}}spec.template.spec.containers[].args{{< /hover >}}
```
Line="7" doesn't seem like the right line here.

jeanduplessis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+section of the deployment.
+
+For example, to change the `sync-interval` add 
+{{<hover label="args" line="10" >}}--sync-interval=30m{{< /hover >}}.
```
Line="10" doesn't seem like the right line here.

plumbis replied:
fixed

jeanduplessis reviewed (changes requested):
@plumbis Thanks for starting this concept doc. I found it very valuable. Thanks for your patience with some of the noob questions on this content.

plumbis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+
+## Crossplane pod
+
+### Init pod
```
Fixed all pod/container references.

plumbis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+periodically poll the resource to find it's state. The default polling rate is
+one minute. Change the polling rate with the `---poll-interval` pod argument.
+
+Managed resources use polling.
```
What would trigger an MR watch? I figured that MR changes would only happen based on external APIs, so it wouldn't be a notification. 

I wasn't able to figure out what is and isn't assigned a watch in source.

bobh66 replied:
The MR controller is by definition "watching" all MRs of that type - it's the same for all Crossplane resource controllers, they all "watch" the specific composite/managed/etc resources that they are responsible for.  The synch interval is presumably intended to account for any "missed events" and verify that the resource state is what it should be regardless of whether the controller missed an event.  (This is my understanding - if Nic tells you something different, believe him.)

plumbis replied:
Insightful! Do you have an example of a MR event that would be caught by a watch?

bobh66 replied:
Probably the simplest case is deletion - when someone or something deletes an MR kubernetes will add the `deletionTimestamp` and `deletionGracePeriodSeconds` attributes to the `metadata` section which triggers an event to the controller, the reconciler sees that the MR has been deleted and calls the external Delete() function to trigger resource deletion via the provider.  Another case would if the someone or something changed a value in the `spec` of the resource - that would also trigger an event to the controller which would detect the difference and trigger an update via the provider.

Note that polling/syncing and watched events are used together - the events are kubernetes-detected changes in (usually desired) state as above, where polling and syncing are looking for drift on the remote side (actual state).  Polling and syncing are time-based (check for differences every 5m/1h) where events are immediate (hey something changed you might care about - or not)

plumbis replied:
Thank you! I've added a MRs to the list of resources that uses watches as well as a notice under "MRs use polling" to explain when an MR would rely on the watch vs poll. 

I'm not trying to get too deep but just create awareness in the user.

plumbis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,231 @@
+interval. 
+
+The most that Crossplane acts on objects is the 
+`--max-reconcile-rate` rate. This rate is 10 times per second by default.
```
Do you mean that it's 10 API calls per second AND 10 goroutines?  Meaning the value is two different settings?

bobh66 replied:
Yes - as shown here:  https://github.com/crossplane/crossplane/blob/master/cmd/crossplane/core/core.go#L178

it is used for both MaxConcurrentReconciles (number of goroutines) and GlobalRateLimiter

plumbis replied:
Thanks again. I added a bunch of stuff here to expand on it. I'm not 100% sure I got it right and accurate so I'd love a second review when you have a chance.

negz commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,321 @@
+[Crossplane Install]({{<ref "../software/install">}}) section. 
+{{< /hint >}}
+
+### Edit the deployment
```
Could we link to upstream Kubernetes documentation for this? This feels like we're teaching folks how to use Kubernetes.

plumbis replied:
The amount of "teaching kubernetes" is light and just provides practical context to applying the settings discussed earlier.

negz reviewed (commented):
I'm seeing a bunch of changes that don't seem to be related to the PR description - i.e. updates to content under the v1.12 and v1.9 directories. Is this intentional?

plumbis commented:
> I'm seeing a bunch of changes that don't seem to be related to the PR description - i.e. updates to content under the v1.12 and v1.9 directories. Is this intentional?

I pulled in the v1.12 release either on accident or purpose. I've rebased and removed that commit. Only the pods.md file and vale updates are in the PR now.

plumbis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,321 @@
+defining webhook configurations. 
+
+The core CRDs installed by the init container include: 
+* Composite Resource Definitions, Compositions, Configurations and Providers
+* `locks` to manage package dependencies
+* `controllerconfigs` to apply settings to installed Providers
+* `storeconfigs` for connecting external secret stores like 
```
Do you mean why are some of them in code ticks?

negz replied:
In part, yes. Everything you've listed on the highlighted lines is both a Crossplane concept and a concrete API type. You could write any of them in three ways:

1. Composite Resource Definitions, Providers, Locks, Store Configs, etc.(Typical prose)
2. `CompositeResourceDefinition`, `Provider`, `Lock`, `StoreConfig`, etc. (`KindName` style)
3. `compositeresourcedefinitions`, `providers` `locks`, `storeconfigs`, etc. (`pluralname` style)

I would recommend using one of the first two styles consistently. The `pluralname` style where a type is written as lower case, plural, and in a code block is very uncommon relative to the `KindName` style.

plumbis replied:
After digging through the [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-upper-camel-case-for-api-objects), the [PR that pushed that guideline](https://github.com/kubernetes/website/pull/23497) and some [related discussion](https://github.com/kubernetes/website/issues/20862) the best answer seems to be the k8s style "[UpperCamelCase](https://en.wikipedia.org/wiki/Camel_case)"

The only problem I have with this is that it's confusing in how k8s describes when to use sentence-style capitalization and their docs are inconsistent in this styling so there aren't great examples to follow.

That being said I've updated this to use UpperCamelCase in the called out section and tried to update any other references in the rest of the doc.

plumbis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,321 @@
+the desired state of Crossplane resources, manages leader elections and process
+webhooks. 
+
+#### Reconcile loop
+
+The core container operates on a _reconcile loop_, constantly checking the 
+status of deployed resources and correcting any "drift." After checking a 
+resource Crossplane waits some time and checks again.
```
I don't think this should be moved, since the page flow is about the pods and their interactions/responsibilities but I do think the [root concepts page](https://docs.crossplane.io/v1.12/concepts/) cover this at or near the top instead of being a rehash of what's in the [Introduction](https://docs.crossplane.io/v1.12/getting-started/introduction/) page and what's spelled out in the individual concept sections.

negz replied:
Yeah, I agree. It's not so much that I don't want it mentioned here insomuch as I think it should appear somewhere more prominent. If that were the case I could see this being a note on this page rather than an entire heading.

plumbis replied:
I understand. Making it just a note doesn't feel like it's enough. I think the solution is we need to give reconciliation more attention in the parent /concepts page.

plumbis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,321 @@
+provider environment.
+{{< /hint >}}
+
+Crossplane checks all resources, both watched and polled resources, to
```
I added a line earlier "Some resources may be both watched and polled"

bobh66 replied:
From the user's perspective it's essentially all resources that they care about/create are both watched and polled.

plumbis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,321 @@
+Reducing the `--max-reconcile-rate`, or making it smaller, increases the
+CPU resources Crossplane uses.
+
+Increasing the `--max-reconcile-rate`, or making it larger, reduces CPU 
+resources Crossplane uses, but increases the amount of time until changed 
+resources are fully synced. 
```
Had those switched. Fixed.

plumbis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,321 @@
+state Crossplane checks the `--max-reconcile-rate` times per second. By default,
+the `--max-reconcile-rate` value of 10 means checking 10 times per second. 
+
+{{< hint "note" >}}
+Kubernetes defines the initial rate of 100 times per second. This isn't
+configurable.
+{{< /hint >}}
```
Gladly willing to change this. It was from trying to understand how the rate limiter works and ended up on Dan's [blog about client-go](https://danielmangum.com/posts/controller-runtime-client-go-rate-limiting/#the-default-controller-rate-limiter) describing the token bucket rate limiter it uses.

plumbis replied:
Removed

plumbis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,321 @@
+default. Use the `--sync-interval` Crossplane pod argument to change this
+interval. 
+
+The `--max-reconcile-rate` rate defines two different settings at the same time:
+
+* The global retry rate for checking failed objects.
+* The number of threads created to reconcile resources.
```
I've attempted to simplify this section.

plumbis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,321 @@
+* creating and applying RBAC policies to Provider Service Accounts, allowing 
+  them to control their managed resources
+* allowing the `crossplane` Service Account to create managed resources
+* creating and applying policies for Crossplane composite resources
```
>It's not clear to me what this means.

What isn't clear? I'm trying to indicate that the RBAC manager creates "some policies" that Crossplane needs to operate.

negz replied:
> What isn't clear? 

I think it's the missing "what for". The four broad things the RBAC manager does is:

1. Create and bind a role to Crossplane's service account, so it can control managed resources.
2. Create and bind roles to provider service accounts, so they can reconcile their managed resources.
3. Create cluster roles that you can bind to grant users access to Crossplane resources in all namespaces.
4. Create roles that you can bind to grant users access to Crossplane resources in a specific namespace.

Noting that 4 is disabled when you run the RBAC manager in `Basic` mode.

At the moment you capture the why of the first two well - these roles enable Crossplane and provider pods do what they need to do. What's throwing me off is that 3 and 4 are a bit over-simplified:

* I would briefly mention. what the roles are for.
* The roles grant access to claims, MRs, secrets, etc in addition to XRs. So I would suggest being non-specific and saying "Crossplane resources". Calling out XRs specifically implies to me they grant only that.

plumbis replied:
I was confused in your earlier comments on the `Roles` vs `ClusterRoles`. I've rewritten the section and I chose not to detail the Roles. Maybe in the future we can create an RBAC section that would go into details.

plumbis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,321 @@
+
+The RBAC manager container manages these functions. 
+
+The RBAC manager pod supports two policy options:
+* `All` - The default policy, configures and manages all RBAC capabilities.
+* `Basic` - Only manage policies related to Composite resources. 
+
+Use the `Basic` policy option when manually assigning RBAC policies to
+Crossplane providers and provider resources. The RBAC manager still creates
+cluster roles for composite resources. 
```
I edited this section to hint at that. I'm trying to avoid fully describing every ClusterRole/Role that's created and the policies around it.

negz replied:
Understood. I think it's worth documenting these policies in detail somewhere, but the "pods" page probably isn't it. Let's discuss in the thread at https://github.com/crossplane/docs/pull/412#discussion_r1183121100.

plumbis replied:
I was confused in your earlier comments on the `Roles` vs `ClusterRoles`. I've rewritten the section and I chose not to detail the `Roles`. Maybe in the future we can create an RBAC section that would go into details.

negz commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,321 @@
+{{< /hint >}}
+##### Reconcile retry rate
+
+When Crossplane checks a resource and it's not in the desired
+state they 100 times per second. If the resource is still not in the desired 
+state Crossplane checks the `--max-reconcile-rate` times per second. By default,
+the `--max-reconcile-rate` value of 10 means checking 10 times per second. 
```
Re-opening this thread as there's still some strange grammar in here, e.g.:

> When Crossplane checks a resource and it's not in the desired state __they__ 100 times per second.

Does my comment above help communicate what `--max-reconcile-rate` does? I'd like to make sure we remove any mention of anything happening 100 times per second - it's wrong and I think it would be worrying to readers. Let me know if I can help better explain how this works. Happy to take a stab at writing this section myself if necessary.

plumbis replied:
I've simplified the wording here and removed the reference to 100x

negz commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,321 @@
+
+The RBAC manager pod supports two policy options:
+* `All` - The default policy, configures and manages all RBAC capabilities.
+* `Basic` - Doesn't enable namespaced RBAC roles.
+
+Using the `Basic` policy option prevents creating namespaced Crossplane RBAC
```
I know you're attempting to be succinct here, but if you could briefly mention what the namespaced Crossplane RBAC roles are for I think it would help contextualize why folks would potentially use `Basic`.

Something like:

> When the RBAC manager uses the `All` policy it will create a series of opinionated RBAC roles in each namespace. You can bind these RBAC roles to users to grant them access to Crossplane types within a namespace.

plumbis commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,321 @@
+status of deployed resources and correcting any "drift." After checking a 
+resource Crossplane waits some time and checks again.
+
+{{<hint "note" >}}
+The Crossplane pod only reconciles core Crossplane components, including claims
+and composite resources. Providers are responsible for reconciling their managed
+resources. 
+{{< /hint >}}
```
moved.

negz commented on content/master/concepts/pods.md:
```diff
@@ -0,0 +1,384 @@
+
+##### Number of reconcilers
+
+The second value `--max-reconcile-rate` defines is the number of threads
```
I'd still prefer not to talk about threads. It's not accurate. Could we find some other approachable way to phrase this? i.e. How many resources can be reconciled concurrently?

plumbis replied:
I'm struggling with this. I understand why you want to avoid `thread` but it seems like we're hinting at something without actually naming it. How is this edit? 
---
The second value `--max-reconcile-rate` defines is the number of
resources that can reconcile at once. If there are more resources than
configured `--max-reconcile-rate` the remaining resources must wait until
Crossplane reconciles a an existing resource.

Read the [Change Pod Settings]({{<ref "#change-pod-settings">}}) section for
instructions on applying these settings.

negz replied:
That sounds great to me.

negz commented on content/master/concepts/pods.md:
````diff
@@ -0,0 +1,384 @@
+View the full RBAC policy with 
+
+```shell
+kubectl describe clusterrole crossplane:aggregate-to-admin
````
```suggestion
kubectl describe clusterrole crossplane-admin
```

`crossplane:aggregate-to-admin` isn't the full role, it's just part of it. `crossplane-admin` is the full role.

(Ditto for the edit, view, and browse roles below.)

negz reviewed (commented):
I've only reviewed the content in the `master` directory, not `v1.11` or `v1.12` which I presume are identical.

negz reviewed (approved):
This looks great, thanks for all the effort!
